paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_cross_op_memory_reuse_pass.cc

```diff
@@ -0,0 +1,498 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "paddle/fluid/framework/details/computation_op_handle.h"
+#include "paddle/fluid/framework/details/multi_devices_helper.h"
+#include "paddle/fluid/framework/details/share_tensor_buffer_op_handle.h"
+#include "paddle/fluid/framework/ir/memory_optimize_pass/memory_optimization_var_info.h"
+#include "paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.h"
+#include "paddle/fluid/framework/ir/memory_optimize_pass/op_graph_view.h"
+#include "paddle/fluid/framework/ir/memory_optimize_pass/reference_count_pass_helper.h"
+#include "paddle/fluid/framework/ir/pass.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+using OpHandleBase = details::OpHandleBase;
+using ComputationOpHandle = details::ComputationOpHandle;
+using VarHandle = details::VarHandle;
+using VarHandleBase = details::VarHandleBase;
+using DummyVarHandle = details::DummyVarHandle;
+
+enum NodeDependency { kSame = 0, kNoDep = 1, kBefore = 2, kAfter = 3 };
+
+static NodeDependency ReverseNodeDependency(NodeDependency dep) {
+  return dep == NodeDependency::kBefore
+             ? NodeDependency::kAfter
+             : (dep == NodeDependency::kAfter ? NodeDependency::kBefore : dep);
+}
+
+class BufferSharedCrossOpMemoryReusePass : public MemoryReusePass {
+ protected:
+  std::string ReuseType() const override { return "cross_op_memory_reuse"; }
```

Comment:
Should we use const value for this kind of string if it may be referred somewhere?

Reply:
This string is only used in logging without referencing in other places. Maybe it is not necessary to use const value.